### ISSUE TITLE
SIM222 overstepping

### ISSUE THREAD

**Issue by martinlehoux** [astral-sh/ruff#21136]
### Summary

Hello!
I was updating to v0.14.2, I had only one unsafe fix with SIM222. I think the rule is overstepping with this example:

```py
def get_items() -> tuple[Item, ...] | None:
    return tuple(item for item in Item.objects.all()) or None
```

suggested fix:
```py
def get_items() -> tuple[Item, ...] | None:
    return tuple(item for item in Item.objects.all())
```

I'm not saying the code has the best design, but the suggestion changes the function behavior (of course not critical, this is an unsafe fix)

By the way, the fix not suggested in this case (no comprehension)
```py
def get_items() -> tuple[Item, ...] | None:
    return tuple(Item.objects.all()) or None
```

**Comment by ntBre** [comment]
Thanks for the report! This does seem incorrect since `tuple(empty_generator())` is falsy, even if the generator itself is truthy.

```pycon
>>> if tuple(x for x in range(0)): print("True")
...
>>> if (x for x in range(0)): print("True")
...
True
```

**Comment by gsakkis** [comment]
I stumbled on this too. One quick note, it doesn't affects just tuples:
```diff
-        print(list(x for x in range(10) if x % 2 == 2) or [42])
+        print([x for x in range(10) if x % 2 == 2])
```

```diff
-        print(set(x for x in range(10) if x % 2 == 2) or {42})
+        print({x for x in range(10) if x % 2 == 2})
```

```diff
-        print(frozenset(x for x in range(10) if x % 2 == 2) or frozenset([42]))
+        print(frozenset(x for x in range(10) if x % 2 == 2))
```
